### ISSUE TITLE
Figure out how to add these helpers

### ISSUE THREAD

**Issue by nwmahoney** [zacclark/Eventual#1]
```
extension Eventual where T : ResultType {
    func mapResult<U>(transform: T.Value -> U) -> Eventual<Result<U, T.Error>> {
        return self.map { $0.map(transform) }
    }

    func flatMapResult<U>(transform: T.Value -> Result<U, T.Error>) -> Eventual<Result<U, T.Error>> {
        return self.map { $0.flatMap(transform) }
    }

    func mapResultError<Error2>(transform: T.Error -> Error2) -> Eventual<Result<T.Value, Error2>> {
        return self.map { $0.mapError(transform) }
    }

    func flatMapResultError<Error2>(transform: T.Error -> Result<T.Value, Error2>) -> Eventual<Result<T.Value, Error2>> {
        return self.map { $0.flatMapError(transform) }
    }
}
```

**Comment by zacclark** [comment]
Updated to use Deferred's generic types:

```
extension FutureType where Value : ResultType {
    typealias ResultValue = Value.Value

    /// Transform the successful value once it exists
    func mapResult<U>(transform: ResultValue -> U) -> Future<Result<U, Value.Error>> {
        return self.map { $0.map(transform) }
    }

    /// Transform the successful value **to another Future value** and re-wrap
    /// This is useful if you need to do more async work only after a success
    func mapResultFuture<U>(transform: ResultValue -> Future<U>) -> Future<Result<U, Value.Error>> {
        return self.flatMap { result -> Future<Result<U, Value.Error>> in
            result.analysis(
                ifSuccess: { val in transform(val).map(Result.Success) },
                ifFailure: { err in Future(value: .Failure(err)) }
            )
        }
    }

    /// Transform the value into another Result and rewrap
    /// This is useful if you have cases where a "success" needs to be turned into an error
    func flatMapResult<U>(transform: ResultValue -> Result<U, Value.Error>) -> Future<Result<U, Value.Error>> {
        return self.map { $0.flatMap(transform) }
    }

    /// Transform the error once it exists
    func mapResultError<Error2>(transform: Value.Error -> Error2) -> Future<Result<ResultValue, Error2>> {
        return self.map { $0.mapError(transform) }
    }

    /// Transform the error into another Result and rewrap
    /// This is useful if you want to "recover" an error back into a success
    func flatMapResultError<Error2>(transform: Value.Error -> Result<Value.Value, Error2>) -> Future<Result<ResultValue, Error2>> {
        return self.map { $0.flatMapError(transform) }
    }
}
```
